base/cvd/cuttlefish/host/libs/web/oauth2_consent.cpp

```diff
@@ -0,0 +1,287 @@
+
+  std::string code;
+
+  if (ssh) {
+    http_server.reset(nullptr);
+
+    std::cout << "Open this URL in your browser: " << consent.rdbuf();
+
+    std::cout << "\n\nThis leads to a 'connection refused' page.\n";
+    std::cout << "Copy and paste that page's URL here: ";
+
+    std::string code_url;
+    std::getline(std::cin, code_url);
+
+    code = CF_EXPECT(CodeFromUrl(code_url));
+  } else {
+    std::cout << "Opening a browser for the consent flow.\n";
+    std::cout << "Using SSH? Please run this command again with `--ssh`.\n";
+
+    CF_EXPECT_EQ(Execute({"/usr/bin/xdg-open", consent.str()}), 0);
+
+    code = CF_EXPECT(http_server->CodeFromClient());
+  }
+
+  // TODO: schuffelen - Deduplicate with `RefreshCredentialSource::Refresh()`
+  std::stringstream exchange;
+  exchange << "code=" << code << "&";
+  exchange << "client_id=" << request.client_id << "&";
+  exchange << "client_secret=" << request.client_secret << "&";
+  exchange << "redirect_uri=" << redirect_uri << "&";
+  exchange << "grant_type=authorization_code";
+
+  constexpr char kExchangeUrl[] = "https://oauth2.googleapis.com/token";
+  std::vector<std::string> headers = {
+      "Content-Type: application/x-www-form-urlencoded"};
+  Json::Value token_json =
+      CF_EXPECT(http_client.PostToJson(kExchangeUrl, exchange.str(), headers))
+          .data;
+
+  CF_EXPECT(!token_json.isMember("error"),
+            "Response had \"error\" but had http success status. Received '"
+                << token_json << "'");
+
+  CF_EXPECTF(token_json.isMember(kScope), "No '{}'", kScope);
+  CF_EXPECT_EQ(token_json[kScope].type(), Json::ValueType::stringValue);
+  std::string response_scope = token_json[kScope].asString();
+  std::vector<std::string> response_scopes = Tokenize(response_scope, " ");
+  for (const std::string& scope : request.scopes) {
+    CF_EXPECTF(Contains(response_scopes, scope), "Response missing '{}'",
+               scope);
+  }
+
+  CF_EXPECTF(token_json.isMember(kRefreshToken), "No '{}'", kRefreshToken);
+  CF_EXPECT_EQ(token_json[kRefreshToken].type(), Json::ValueType::stringValue);
+
+  return token_json[kRefreshToken].asString();
+}
+
+static constexpr char kCredentials[] = "credentials";
+
+Result<std::unique_ptr<CredentialSource>> Oauth2Login(
+    HttpClient& http_client, const Oauth2ConsentRequest& request, bool ssh) {
+  std::string refresh_token =
+      CF_EXPECT(GetRefreshToken(http_client, request, ssh));
+
+  Json::Value serialized;
+  serialized[kClientId] = request.client_id;
+  serialized[kClientSecret] = request.client_secret;
+  serialized[kRefreshToken] = refresh_token;
+  for (const std::string& scope : request.scopes) {
+    serialized[kScope].append(scope);
+  }
+
+  uint32_t checksum = ScopeChecksum(request.scopes);
+  std::string filename = fmt::format("{}/{}.json", kCredentials, checksum);
+  std::string contents = serialized.toStyledString();
+
+  CF_EXPECT(WriteCvdDataFile(filename, std::move(contents)));
```

Comment:
`RefreshCredentialSource` handles multiple existing file formats, some of which do store an access token and expiration time. `RefreshCredentialSource` isn't an authority on any of the prior file formats so it doesn't modify the files to avoid corrupting them.

Now that `cvd` is the authority on at least one file format, persisting the access token does seem like a reasonable feature request to consider in the future. I've filed b/376321455 to track persisting the access token.